### DISCUSSION TITLE
Treeview

### DISCUSSION THREAD

**Discussion by HlDANN** [TomSchimansky/CustomTkinter#524] [Ideas]
Hi guys!! 
I'm using this lib in a private project and I'm really enjoying it. Do we have treeview implemented or forecast to implement in the future? :D

**Comment by Akascape** [comment]
Custom treeview is not currently implemented.

But you can use the normal tkinter treeview with customisations. 

## Here is the example:

![Demo](https://user-images.githubusercontent.com/89206401/196231434-9fae137b-0139-46b3-87a1-582b1bb28fa5.jpg)

## Code
```python
import tkinter
from tkinter import ttk
import customtkinter

customtkinter.set_appearance_mode("Dark")
customtkinter.set_default_color_theme("blue")

root = customtkinter.CTk()
root.geometry("250x300")
root.title("Treeview x Customtkinter")

frame_1 = customtkinter.CTkFrame(master=root)
frame_1.pack(pady=20, padx=20, fill="both", expand=True)

label = customtkinter.CTkLabel(master=frame_1,text="Treeview")
label.grid(pady=10)

###Treeview Customisation (theme colors are selected)
bg_color = root._apply_appearance_mode(customtkinter.ThemeManager.theme["CTkFrame"]["fg_color"])
text_color = root._apply_appearance_mode(customtkinter.ThemeManager.theme["CTkLabel"]["text_color"])
selected_color = root._apply_appearance_mode(customtkinter.ThemeManager.theme["CTkButton"]["fg_color"])

treestyle = ttk.Style()
treestyle.theme_use('default')
treestyle.configure("Treeview", background=bg_color, foreground=text_color, fieldbackground=bg_color, borderwidth=0)
treestyle.map('Treeview', background=[('selected', bg_color)], foreground=[('selected', selected_color)])
root.bind("<<TreeviewSelect>>", lambda event: root.focus_set())

##Treeview widget data
treeview = ttk.Treeview(frame_1, height=6, show="tree")
treeview.grid(padx=10)
treeview.insert('', '0', 'i1', text ='Python')
treeview.insert('', '1', 'i2', text ='Customtkinter')
treeview.insert('', '2', 'i3', text ='Tkinter')
treeview.insert('i2', 'end', 'Frame', text ='Frame')
treeview.insert('i2', 'end', 'Label', text ='Label')
treeview.insert('i3', 'end', 'Treeview', text ='Treeview')
treeview.move('i2', 'i1', 'end')
treeview.move('i3', 'i1', 'end')

root.mainloop()
```
If you are looking for a table widget, here is one: https://github.com/Akascape/CTkTable

**Reply by HlDANN** [reply]
Understood! Thanks

**Reply by 0xHouss** [reply]
i'm working on making a custom treeview but from scratch if someone wanna help me here my discord `JL Houss#8625`

**Comment by Mank0o** [comment]
Hi Akascape,

thank you for sharing the code but when I run it I get this error
I'm using customtkinter-5.0.2

```
Traceback (most recent call last):
  File "c:\Users\Mankoo\OneDrive\RGR Trucking\1 Not sent\Code\server\Treeview_customtkinker.py", line 21, in <module>
    treestyle.configure("Treeview", background=customtkinter.ThemeManager.theme["color"]["frame_low"][1],
KeyError: 'color
```

**Reply by Akascape** [reply]
@Mank0o I wrote this code before the version 5.0.0. The theme structure was changed after the update due to which it was showing this error.
But I have updated the code. Now you can try it.

**Reply by Mank0o** [reply]
Thank you @Akascape

**Comment by B16f00t** [comment]
When I try to switch from dark mode to light mode, the treeview appearance is black, when it should be white, would it be possible to make it vary depending on the mode selected?

**Reply by Akascape** [reply]
Configure the treeview again, just change the last color theme from [1] to [0]. Example-

Dark:
`background=customtkinter.ThemeManager.theme["CTkFrame"]["fg_color"][1] `

Light:
`background=customtkinter.ThemeManager.theme["CTkFrame"]["fg_color"][0]`

**Comment by B16f00t** [comment]
Yes I know that, the problem is that in the app I can change the appearance in execution and I don't know what event triggers set_appearance_mode.
I would like to be able to set in dark or light, the treeview according to choose one or the other in execution.

**Reply by Akascape** [reply]
@B16f00t there is event which can tell you the current appearance mode:

```python
if customtkinter.get_appearance_mode()=="Dark":
    x=1
else:
    x=0

#Now you can get the correct color
background = customtkinter.ThemeManager.theme["CTkFrame"]["fg_color"][x]
```

**Reply by TomSchimansky** [reply]
It's not actually an event. You can register a callback for the appearance mode change with:
```python
customtkinter.AppearanceModeTracker.add(callback)
```
The callback will get a string argument being either "Light" or "Dark".

**Comment by B16f00t** [comment]
That's great, I was able to solve the problem with customtkinter.AppearanceModeTracker.add(callback).
Thank you very much.

**Comment by Neofare** [comment]
how do you add internal padding inside the entry widget?

**Reply by Flivo254** [reply]
To add internal padding use ipadx=(px) on x-axis and ipady=(px) on y-axis

**Comment by adamtcdev** [comment]
![image](https://github.com/TomSchimansky/CustomTkinter/assets/157731913/56645a59-a7d6-45fb-ad3e-7673a9f06bcf)
I don't think it's working properly. Anyone help?

**Comment by Sander-Roni** [comment]
Akascape e se for pra criar tipo um explorador de arquivos?, ex... com a biblioteca OS?, no sentido de exibir mais do que os arquivos que você colocou ali.
